src/query.rs

```diff
@@ -1783,11 +1789,92 @@ fn release_borrow<Q: Query>(archetypes: &[Archetype]) {
     }
 }
 
+#[cfg(feature = "std")]
+pub(crate) type QueryCache = RwLock<TypeIdMap<Arc<dyn Any + Send + Sync>>>;
```

Comment:
If these locks are short-lived and this is the only reason to require `std`, I would still suggest looking into using an "atomic cell" or some other form of borrow checking, especially if reentracy is controlled by this not being public API.

Reply:
How could we gracefully expand a `TypeIdMap` inside an atomic cell? Inserting requires unique access.

Reply:
Using an API like [`AtomicRefCell`](https://docs.rs/atomic_refcell/latest/atomic_refcell/struct.AtomicRefCell.html), i.e. `RefCell` implemented using atomics. I use it extensively to ensure cooperatively scheduled futures (which do not need blocking) remain `Send`.

Reply:
From the documentation of that crate:
> for consumers that can guarantee that they will never mutably borrow the contents concurrently with immutable borrows, an RwLock is overkill

We cannot make such a guarantee.

Reply:
Propagating that requirement to callers would prevent overlapping read-only queries, which is a major regression.

Reply:
You are right, you still need exclusive access for compatible queries to initialise the cache. Sorry for the noise!